File: backend/src/main/kotlin/com/liyaqa/shared/domain/model/PrayerTime.kt
package com.liyaqa.shared.domain.model

import java.time.LocalDate
import java.time.LocalTime

/**
 * Prayer calculation methods supported for prayer time calculation.
 * UMM_AL_QURA is the official Saudi method.
 */
enum class PrayerCalculationMethod {
    /** Umm Al-Qura University, Makkah - Official Saudi Arabia method */
    UMM_AL_QURA,
    /** Muslim World League */
    MUSLIM_WORLD_LEAGUE,
    /** Egyptian General Authority of Survey */
    EGYPTIAN,
    /** University of Islamic Sciences, Karachi */
    KARACHI,
    /** Islamic Society of North America */
    ISNA,
    /** Union of Islamic Organizations of France */
    UOIF,
    /** Dubai - UAE */
    DUBAI,
    /** Qatar */
    QATAR,
    /** Kuwait */
    KUWAIT,
    /** Singapore */
    SINGAPORE
}

/**
 * Represents prayer times for a specific date and location.
 */
data class PrayerTime(
    val date: LocalDate,
    val fajr: LocalTime,
    val sunrise: LocalTime,
    val dhuhr: LocalTime,
    val asr: LocalTime,
    val maghrib: LocalTime,
    val isha: LocalTime
) {
    /**
     * Returns the next prayer time from the given current time.
     */
    fun getNextPrayer(currentTime: LocalTime): NextPrayer {
        return when {
            currentTime.isBefore(fajr) -> NextPrayer("FAJR", fajr)
            currentTime.isBefore(sunrise) -> NextPrayer("SUNRISE", sunrise)
            currentTime.isBefore(dhuhr) -> NextPrayer("DHUHR", dhuhr)
            currentTime.isBefore(asr) -> NextPrayer("ASR", asr)
            currentTime.isBefore(maghrib) -> NextPrayer("MAGHRIB", maghrib)
            currentTime.isBefore(isha) -> NextPrayer("ISHA", isha)
            else -> NextPrayer("FAJR", fajr) // Next day's Fajr
        }
    }

    /**
     * Checks if the given time is during prayer time.
     * @param currentTime The time to check
     * @param bufferMinutes Minutes before and after prayer to consider as "during prayer"
     */
    fun isDuringPrayerTime(currentTime: LocalTime, bufferMinutes: Int = 30): Boolean {
        val prayers = listOf(fajr, dhuhr, asr, maghrib, isha)
        return prayers.any { prayerTime ->
            val start = prayerTime.minusMinutes(5)
            val end = prayerTime.plusMinutes(bufferMinutes.toLong())
            !currentTime.isBefore(start) && !currentTime.isAfter(end)
        }
    }

    /**
     * Returns the current prayer period or null if not during prayer.
     */
    fun getCurrentPrayerPeriod(currentTime: LocalTime, bufferMinutes: Int = 30): String? {
        val prayers = listOf(
            "FAJR" to fajr,
            "DHUHR" to dhuhr,
            "ASR" to asr,
            "MAGHRIB" to maghrib,
            "ISHA" to isha
        )

        for ((name, prayerTime) in prayers) {
            val start = prayerTime.minusMinutes(5)
            val end = prayerTime.plusMinutes(bufferMinutes.toLong())
            if (!currentTime.isBefore(start) && !currentTime.isAfter(end)) {
                return name
            }
        }
        return null
    }
}

/**
 * Represents the next upcoming prayer.
 */
data class NextPrayer(
    val name: String,
    val time: LocalTime
)

/**
 * Prayer settings embedded in Club entity.
 */
data class PrayerSettings(
    val city: String?,
    val latitude: Double?,
    val longitude: Double?,
    val calculationMethod: PrayerCalculationMethod,
    val bufferMinutes: Int,
    val blockCheckinDuringPrayer: Boolean
) {
    fun isConfigured(): Boolean = latitude != null && longitude != null
}

File: frontend/shared/src/types/prayer-time.ts
/**
 * Prayer Time types for Saudi Arabia market features.
 */

// ==================== ENUMS ====================

export type PrayerCalculationMethod =
  | "UMM_AL_QURA"
  | "MUSLIM_WORLD_LEAGUE"
  | "EGYPTIAN"
  | "KARACHI"
  | "ISNA"
  | "UOIF"
  | "DUBAI"
  | "QATAR"
  | "KUWAIT"
  | "SINGAPORE";

export type PrayerName = "FAJR" | "SUNRISE" | "DHUHR" | "ASR" | "MAGHRIB" | "ISHA";

// ==================== PRAYER TIME RESPONSES ====================

export interface PrayerTimeResponse {
  date: string;
  fajr: string;
  sunrise: string;
  dhuhr: string;
  asr: string;
  maghrib: string;
  isha: string;
}

export interface NextPrayerResponse {
  name: PrayerName;
  time: string;
}

export interface CheckInStatusResponse {
  blocked: boolean;
  currentPrayer: PrayerName | null;
  messageEn: string | null;
  messageAr: string | null;
}

export interface PrayerSettingsResponse {
  city: string | null;
  latitude: number | null;
  longitude: number | null;
  calculationMethod: PrayerCalculationMethod;
  bufferMinutes: number;
  blockCheckinDuringPrayer: boolean;
  isConfigured: boolean;
}

// ==================== CITY INFO ====================

export interface CityInfo {
  name: string;
  nameAr: string;
  latitude: number;
  longitude: number;
}

export interface SupportedCitiesResponse {
  cities: CityInfo[];
}

// ==================== CALCULATION METHOD INFO ====================

export interface CalculationMethodInfo {
  code: PrayerCalculationMethod;
  nameEn: string;
  nameAr: string;
  description: string;
}

// ==================== UPDATE REQUEST ====================

export interface UpdatePrayerSettingsRequest {
  city?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  calculationMethod?: PrayerCalculationMethod | null;
  bufferMinutes?: number | null;
  blockCheckinDuringPrayer?: boolean | null;
}

// ==================== HELPER TYPES ====================

export interface PrayerTimeItem {
  name: PrayerName;
  nameEn: string;
  nameAr: string;
  time: string;
  isPast: boolean;
  isNext: boolean;
}

// Prayer name translations
export const PRAYER_NAMES: Record<PrayerName, { en: string; ar: string }> = {
  FAJR: { en: "Fajr", ar: "الفجر" },
  SUNRISE: { en: "Sunrise", ar: "الشروق" },
  DHUHR: { en: "Dhuhr", ar: "الظهر" },
  ASR: { en: "Asr", ar: "العصر" },
  MAGHRIB: { en: "Maghrib", ar: "المغرب" },
  ISHA: { en: "Isha", ar: "العشاء" },
};

// Calculation method translations
export const CALCULATION_METHODS: Record<
  PrayerCalculationMethod,
  { en: string; ar: string; description: string }
> = {
  UMM_AL_QURA: {
    en: "Umm Al-Qura (Saudi Arabia)",
    ar: "أم القرى (السعودية)",
    description: "Official method used in Saudi Arabia",
  },
  MUSLIM_WORLD_LEAGUE: {
    en: "Muslim World League",
    ar: "رابطة العالم الإسلامي",
    description: "Used in Europe, Far East, parts of US",
  },
  EGYPTIAN: {
    en: "Egyptian General Authority",
    ar: "الهيئة المصرية العامة للمساحة",
    description: "Used in Africa, Syria, Lebanon, Malaysia",
  },
  KARACHI: {
    en: "University of Karachi",
    ar: "جامعة كراتشي",
    description: "Used in Pakistan, Bangladesh, India, Afghanistan",
  },
  ISNA: {
    en: "Islamic Society of North America",
    ar: "الجمعية الإسلامية لأمريكا الشمالية",
    description: "Used in North America",
  },
  UOIF: {
    en: "Union of Islamic Orgs (France)",
    ar: "اتحاد المنظمات الإسلامية (فرنسا)",
    description: "Used in France",
  },
  DUBAI: {
    en: "Dubai",
    ar: "دبي",
    description: "Used in UAE",
  },
  QATAR: {
    en: "Qatar",
    ar: "قطر",
    description: "Used in Qatar",
  },
  KUWAIT: {
    en: "Kuwait",
    ar: "الكويت",
    description: "Used in Kuwait",
  },
  SINGAPORE: {
    en: "Singapore",
    ar: "سنغافورة",
    description: "Used in Singapore",
  },
};
